src/Database/Connection.php

```diff
@@ -956,11 +963,26 @@ public function getLogger()
     {
         if ($this->_logger === null) {
             $this->_logger = new QueryLogger();
+            $this->_logger->setConnectionId($this->getConnectionId());
```

Comment:
This is a very specific solution to a more general problem of enriching query logs with additional metadata.

Could we generalize this with a method like `addMetadata()` that takes a key:value array of data that is added to emitted logs? This might also let us refactoring how query timing is collected and included in logs.

Connection configuration could define a closure that prepares the logging metadata the first time it is used. That would let us avoid adding methods to Driver that have only one implementation.

Reply:
Do you mean `$this->_logger->addMetadata(['connection_id' => $this->getConnectionId()]);`?

Reply:
That would be possible but in my opinion that is a wrong thing to do. Usually when an array of items is passed as a method parameter (or a method has many parameters that is practically the same) that is considered as code smell in OOP and refactoring guru M.Fowler suggests to use [Introduce Parameter Object](https://refactoring.guru/introduce-parameter-object) pattern instead. I think it is better for an object to have a list of basic type properties than one array type property (which would contain a list of mixed type elements - int, string, etc.).

Reply:
I don't see how a key value list of logging attributes creates a smell. It would be intended as a way for applications to add application relevant attributes to query logs. The types could be constrained to `array{string: string}`.

Reply:
`$context` argument of `LoggerInterface::debug()` is also an array :wink:. (In 4.x `QueryLogger` is an `LoggerInterface` instance.)

Reply:
Whenever I see an associative array passed as a parameter that smells bad. Sadly but CakePHP has a lot of that. I believe that can be called [data clump](https://en.wikipedia.org/wiki/Data_clump). It is much better to use objects instead. The benefits are briefly explained in [this](https://stackoverflow.com/a/3862899/316041) post.

Reply:
This isn't the PR to discuss overhauling cake's options architecture.

However, some of what you're talking about is due to the language itself some of it is historical.   One of the things that php lacks is nested classes.

The options aren't really data clumps per say.  They are structured options which aren't passed around.  It's similar to creating a function in javascript.

However, insisting that cake be re-designed to avoid aligning your changes with current convention is just going to result in a stale PR.

Reply:
It's not a big deal to change that as asked above and I will do that. But I really have difficulties to understand why would you consider `$this->_logger->addMetadata(['connection_id' => $this->getConnectionId()]);` to be better than `$this->_logger->setConnectionId($this->getConnectionId());`. 
1. It allows typing errors like `$this->_logger->addMetadata(['conection_id' => $this->getConnectionId()]);` that are quite difficult to debug,
2. It allows to set keys that are not handled by the logger class like `$this->_logger->addMetadata(['useless' => true]);`,
3. It does not allow IDEs to autocomplete the name of the array key. There is no such problem with `setConnectionId()` implementation.
4. It will not allow to enforce a different data type for each of array elements, f.ex. 'connection_id' to be int and 'duration' to be a float. There would be no such problem with `function setConnectionId(int $connectionId)` in 4.x.

Reply:
For the `addMedata` method I proposed, mistyped keys would still be included in the log message as log messages would format all metadata keys into emitted messages.

I'm not sure the type safety is required as all the values will be coerced into strings when added to log messages.